plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -457,36 +460,61 @@ public DeltaLakeMetadata(
         this.allowManagedTableRename = allowManagedTableRename;
     }
 
-    public TableSnapshot getSnapshot(ConnectorSession session, SchemaTableName table, String tableLocation, long atVersion)
+    public TableSnapshot getSnapshot(ConnectorSession session, SchemaTableName table, String tableLocation, Optional<Long> atVersion)
     {
-        return getSnapshot(session, table, tableLocation, Optional.of(atVersion));
-    }
+        if (atVersion.isEmpty()) {
+            atVersion = Optional.ofNullable(queriedVersions.get(new QueriedTable(table, Optional.empty())));
+        }
 
-    @VisibleForTesting
-    protected TableSnapshot getSnapshot(ConnectorSession session, SchemaTableName table, String tableLocation, Optional<Long> atVersion)
-    {
-        try {
-            if (atVersion.isEmpty()) {
-                atVersion = Optional.ofNullable(queriedVersions.get(table));
-            }
-            if (atVersion.isPresent()) {
-                long version = atVersion.get();
-                TableSnapshot snapshot = queriedSnapshots.get(new QueriedTable(table, version));
-                checkState(snapshot != null, "No previously loaded snapshot found for query %s, table %s [%s] at version %s", session.getQueryId(), table, tableLocation, version);
-                return snapshot;
-            }
+        QueriedTable queriedTable = new QueriedTable(table, atVersion);
+        if (queriedSnapshots.containsKey(queriedTable)) {
+            return queriedSnapshots.get(queriedTable);
+        }
 
-            TableSnapshot snapshot = transactionLogAccess.loadSnapshot(session, table, tableLocation);
+        try {
+            TableSnapshot snapshot = transactionLogAccess.loadSnapshot(session, table, tableLocation, atVersion);
             // Lack of concurrency for given query is currently guaranteed by DeltaLakeMetadata
-            checkState(queriedVersions.put(table, snapshot.getVersion()) == null, "queriedLocations changed concurrently for %s", table);
-            queriedSnapshots.put(new QueriedTable(table, snapshot.getVersion()), snapshot);
+            checkState(queriedVersions.put(queriedTable, snapshot.getVersion()) == null, "queriedVersions changed concurrently for %s", table);
+            queriedSnapshots.put(new QueriedTable(table, Optional.of(snapshot.getVersion())), snapshot);
```

Comment:
In case that the `queriedTable` is missing a concrete version, do add 

```
checkState(queriedVersions.put(new QueriedTable(table, Optional.of(snapshot.getVersion())), snapshot.getVersion()) == null, "queriedVersions changed concurrently for %s", table);
```

Reply:
The change doesn't work. The condition throw an exception even if it's not changed concurrently.